### PR TITLE
ref(pkg/injector): use XDS structs for envoy bootstrap config

### PR DESCRIPTION
**Pull request by jaellio**
**Description**:

Updates the envoy bootstrap config to be constructed using XDS
structs directly instead of maps. The XDS objects are marshaled
to JSON and then to YAML. Using the XDS structs directly will help
ensure the objects are configured correctly and all required
fields are set. Additionally, this change will make it easier to
enforce the use of the same API version across the injector and
controller.

**Affected area**:
| Functional Area            |     |
| -------------------------- | --- |
| New Functionality          | [ ] |
| Documentation              | [ ] |
| Install                    | [ ] |
| CLI Tool                   | [ ] |
| Certificate Management     | [ ] |
| Control Plane              | [ ] |
| Ingress                    | [ ] |
| Egress                     | [ ] |
| Networking                 | [ ] |
| Observability              | [ ] |
| SMI Policy                 | [ ] |
| Sidecar Injection          | [ X ] |
| Security                   | [ ] |
| Upgrade                    | [ ] |
| Tests                      | [ ] |
| CI System                  | [ ] |
| Demo                       | [ ] |
| Performance                | [ ] |
| Other                      | [ ] |

Please answer the following questions with yes/no.

1. Does this change contain code from or inspired by another project? No
    -   Did you notify the maintainers and provide attribution?

1. Is this a breaking change? No

### CODE REVIEW

File: pkg/injector/envoy_config.go [hunk 1]
```diff
@@ -214,5 +278,37 @@ func getXdsCluster(config envoyBootstrapConfigMeta) map[string]interface{} {
...
+	var jsonObj interface{}
+	// We are using yaml.Unmarshal here (instead of json.Unmarshal) because the
+	// Go JSON library doesn't try to pick the right number type (int, float,
+	// etc.) when unmarshalling to interface{}, it just picks float64
+	// universally. go-yaml does go through the effort of picking the right
+	// number type, so we can preserve number type throughout this process.
+	err = yaml.Unmarshal([]byte(configJSON), &jsonObj)
+	if err != nil {
+		return nil, err
+	}
```
Comment (shashankram) [inline pkg/injector/envoy_config.go:304]: This can be a separate function that references https://github.com/ghodss/yaml/blob/v1.0.0/yaml.go#L89 (references code for a release tag instead of master).

File: pkg/injector/envoy_config_health_probes.go [hunk 2]
```diff
@@ -63,100 +86,229 @@ func getProbeCluster(clusterName string, port int32) map[string]interface{} {
...
 				},
-				"http_filters": []map[string]interface{}{
-					{
-						"name": "envoy.filters.http.router",
+			},
+		}
+		pbHTTPConnectionManager, err := ptypes.MarshalAny(httpConnectionManager)
+		if err != nil {
+			log.Error().Err(err).Msgf("Error marshaling HttpConnectionManager struct into an anypb.Any message")
+			return nil
```
Comment (shashankram) [inline pkg/injector/envoy_config_health_probes.go:132]: nit: have this function return an error else health probes functionality could be broken since an `append()` with `nil` value is valid.

File: pkg/injector/envoy_config.go [hunk 3]
```diff
@@ -214,5 +278,37 @@ func getXdsCluster(config envoyBootstrapConfigMeta) map[string]interface{} {
 				},
 			},
 		},
+	}, nil
+}
+
+func xdsToYaml(m protoreflect.ProtoMessage) ([]byte, error) {
```
Comment (shashankram) [inline pkg/injector/envoy_config.go:284]:
```suggestion
func protoToYaml(m protoreflect.ProtoMessage) ([]byte, error) {
```

File: pkg/injector/envoy_config_health_probes.go [hunk 4]
```diff
@@ -63,100 +86,229 @@ func getProbeCluster(clusterName string, port int32) map[string]interface{} {
...
+						},
 					},
 				},
 			},
 		},
 	}
 }
 
+// GetAccessLog creates an Envoy AccessLog struct.
+func GetAccessLog() []*xds_accesslog_filter.AccessLog {
```
Comment (shashankram) [inline pkg/injector/envoy_config_health_probes.go:216]: un-export this function.

File: pkg/injector/envoy_config_health_probes.go [hunk 5]
```diff
@@ -63,100 +86,229 @@ func getProbeCluster(clusterName string, port int32) map[string]interface{} {
...
+	return []*xds_accesslog_filter.AccessLog{{
+		Name: wellknown.FileAccessLog,
+		ConfigType: &xds_accesslog_filter.AccessLog_TypedConfig{
+			TypedConfig: accessLog,
+		}},
+	}
+}
+
+// GetTCPAccessLog creates an Envoy AccessLog struct.
+func GetTCPAccessLog() []*xds_accesslog_filter.AccessLog {
```
Comment (shashankram) [inline pkg/injector/envoy_config_health_probes.go:231]: un-export this.

File: pkg/injector/test_fixtures/expected_envoy_bootstrap_config.yaml [hunk 6]
```diff
@@ -23,7 +23,7 @@ dynamic_resources:
     resource_api_version: V3
 static_resources:
   clusters:
-  - connect_timeout: 0.25s
+  - connect_timeout: 0.250s
```
Comment (shashankram) [inline pkg/injector/test_fixtures/expected_envoy_bootstrap_config.yaml:26]: Was this change intentional?

File: pkg/injector/envoy_config.go [hunk 7]
```diff
@@ -152,60 +183,93 @@ func (wh *mutatingWebhook) createEnvoyBootstrapConfig(name, namespace, osmNamesp
...
+	}
+	pbUpstreamTLSContext, err := ptypes.MarshalAny(upstreamTLSContext)
+	if err != nil {
+		log.Error().Err(err).Msgf("Error marshaling UpstreamTlsContext struct into an anypb.Any message")
+		return nil, err
+	}
+
+	return &xds_cluster.Cluster{
+		Name:           config.XDSClusterName,
+		ConnectTimeout: durationpb.New(time.Millisecond * 250),
```
Comment (shashankram) [inline pkg/injector/envoy_config.go:242]: This was previously 25ms.

File: pkg/injector/test_fixtures/expected_envoy_bootstrap_config.yaml [hunk 8]
```diff
@@ -72,7 +71,6 @@ static_resources:
     name: liveness_cluster
     type: STATIC
   - connect_timeout: 1s
-    lb_policy: ROUND_ROBIN
```
Comment (shashankram) [inline pkg/injector/test_fixtures/expected_envoy_bootstrap_config.yaml:75]: We should be setting this in the XDS struct.

**Review by shashankram — COMMENTED** [review]
Looks almost good to go. Noticed some discrepancies in values of certain object fields as noted in the comments inline.

File: pkg/injector/envoy_config_health_probes.go (hunk 4 above)
Comment (jaellio) [reply pkg/injector/envoy_config_health_probes.go]: Done

File: pkg/injector/envoy_config.go (hunk 7 above)
Comment (jaellio) [reply pkg/injector/envoy_config.go:242]: I think it was .25 seconds which is 250 ms

File: pkg/injector/test_fixtures/expected_envoy_bootstrap_config.yaml (hunk 6 above)
Comment (jaellio) [reply pkg/injector/test_fixtures/expected_envoy_bootstrap_config.yaml:26]: When the xds structs are marshaled to YAML the 0 after .25 remains

File: pkg/injector/envoy_config.go (hunk 1 above)
Comment (jaellio) [reply pkg/injector/envoy_config.go]: Updated

File: pkg/injector/test_fixtures/expected_envoy_bootstrap_config.yaml (hunk 8 above)
Comment (jaellio) [reply pkg/injector/test_fixtures/expected_envoy_bootstrap_config.yaml:75]: Round robin is the default lb_policy. I am currently setting the lb_policy 'LbPolicy: xds_cluster.Cluster_ROUND_ROBIN', but the value of 'xds_cluster.Cluster_ROUND_ROBIN' is 0. The protojson marshaler considers 0 an empty value and it is therefore omitted. I have not found a way to keep default values. I have the same issue with setting 'codec_type: AUTO' for a filter.

File: pkg/injector/envoy_config_health_probes.go (hunk 2 above)
Comment (jaellio) [reply pkg/injector/envoy_config_health_probes.go:136]: Should the getters also return an error to be handled in `getStaticResources` in `envoy_config.go`?

File: pkg/injector/test_fixtures/expected_envoy_bootstrap_config.yaml (hunk 8 above)
Comment (jaellio) [reply pkg/injector/test_fixtures/expected_envoy_bootstrap_config.yaml:75]: Here is the issue discussing protojson support for masking what fields are omitted https://github.com/golang/protobuf/issues/1183

File: pkg/injector/envoy_config.go [hunk 9]
```diff
@@ -2,81 +2,117 @@ package injector
...
+	"github.com/golang/protobuf/ptypes"
+	"github.com/golang/protobuf/ptypes/any"
+
 	"github.com/openservicemesh/osm/pkg/certificate"
 	"github.com/openservicemesh/osm/pkg/configurator"
 	"github.com/openservicemesh/osm/pkg/constants"
 	"github.com/openservicemesh/osm/pkg/version"
 )
 
 func getEnvoyConfigYAML(config envoyBootstrapConfigMeta, cfg configurator.Configurator) ([]byte, error) {
```
Comment (allenlsy) [inline pkg/injector/envoy_config.go:33]: We encountered similar issue before that, people assumes this methods returns a YAML format string, while it is a byte array.

Do you want to make it return string instead?

E.g. https://github.com/openservicemesh/osm/pull/3388

File: pkg/injector/envoy_config_health_probes.go (hunk 2 above)
Comment (shashankram) [reply pkg/injector/envoy_config_health_probes.go:136]: I would suggest returning an error from functions that can generate errors. In some cases such as this one where `MarshalAny` won't result in an error (because of correctness in the http connection manager struct), it is more about correctness.

File: pkg/injector/test_fixtures/expected_envoy_bootstrap_config.yaml (hunk 6 above)
Comment (shashankram) [reply pkg/injector/test_fixtures/expected_envoy_bootstrap_config.yaml:26]: My bad, this looks good!

File: pkg/injector/envoy_config.go (hunk 7 above)
Comment (shashankram) [reply pkg/injector/envoy_config.go:242]: My bad, looks correct.

**Review by shashankram — APPROVED** [review]

**Review by snehachhabria — APPROVED** [review]

File: pkg/injector/envoy_config.go (hunk 9 above)
Comment (jaellio) [reply pkg/injector/envoy_config.go:33]: Good idea. I will fix this in a follow-up PR
